src/langchain_google_alloydb_pg/utils/pgvector_migrator.py

```diff
@@ -0,0 +1,440 @@
+            batch = []
+
+    if batch:  # Yield remaining items if any
+        yield batch
+
+
+async def _amigrate_pgvector_collection(
+    engine: AlloyDBEngine,
+    collection_name: str,
+    embeddings_service: Embeddings,
+    metadata_columns: Optional[List[str]] = [],
+    destination_table: Optional[str] = None,
+    use_json_metadata: Optional[bool] = False,
+    delete_pg_collection: Optional[bool] = False,
+    insert_batch_size: int = 1000,
+) -> None:
+    """
+    Migrate all data present in a PGVector collection to use separate tables for each collection.
+    The new data format is compatible with the AlloyDB interface.
+
+    Args:
+        engine (AlloyDBEngine): The AlloyDB engine corresponding to the Database.
+        collection_name (str): The collection to migrate.
+        embeddings_service (Embeddings): The embeddings service used in the vectorstore.
+            Embeddings are directly copied from the PGVector database. This service is used only to initialise the vector store.
+        metadata_columns (List[str]): The metadata columns to be created to keep the data in a row-column format.
+            Optional.
+        destination_table (str): The name of the table to insert the data in.
+            Optional. defaults to collection_name.
+        use_json_metadata (bool): An option to keep the PGVector metadata as json in the AlloyDB table.
+            Default: False. Optional.
+        delete_pg_collection (bool): An option to delete the original data upon migration.
+            Default: False. Optional.
+        insert_batch_size (int): Number of rows to insert at once in the table.
+            Default: 1000.
+    """
+    if not use_json_metadata and not metadata_columns:
+        raise ValueError(
+            "Please specify the columns for the new table. "
+            "To store data in JSON format, set use_json_metadata=True when calling this method."
+        )
+
+    if not destination_table:
+        warnings.warn(
+            f"Destination table not set. Destination table would default to {collection_name}. "
+            "Please make sure that there is an existing table with the same name."
+        )
+        destination_table = collection_name
+
+    # Extract data from the collection and batch insert into the new table
+    collection_data = _aextract_pgvector_collection(engine, collection_name)
+    data_batches = _batch_data(collection_data, insert_batch_size)
+
+    if metadata_columns:
+        vector_store = await AsyncAlloyDBVectorStore.create(
+            engine=engine,
+            embedding_service=embeddings_service,
+            table_name=destination_table,
+            metadata_columns=metadata_columns,
+        )
+    else:
+        vector_store = await AsyncAlloyDBVectorStore.create(
+            engine=engine,
+            embedding_service=embeddings_service,
+            table_name=destination_table,
+        )
+
+    tasks = [
+        asyncio.create_task(
```

Comment:
This is where we may need a Python Async Expert. If we have 100k records with a batch size of 1k then we'd create 100 batches, where gather is trying to run them all concurrently. This may slow down the app rather than speed it up. We may want to use "wait()" with "return_when=​FIRST_COMPLETED" then we could run a set of batch concurrently and once they complete we could add more to the batch. Examples https://death.andgravity.com/limit-concurrency#asyncio-wait and https://gist.github.com/kurtisvg/8021ebfa6d6edea0d8b9a28e1efcbfa1.

Reply:
Example code to try:
If we don't have to calculate batch size
```python
pending = []
max_concurrency = 5
async for batch in batches:
   pending.append(asyncio.ensure_future(_ainsert_single_batch(...)) # or in the future vs.add_embeddings(...) instead of insert method
    if len(pending) == max_concurrency:
        done, pending = await asyncio.wait(
            pending, return_when=asyncio.FIRST_COMPLETED
        )
await asyncio.wait(pending)
```

If we do have to calculate batch size
```python
pending = []
max_concurrency = 5
async for batch in batches:
  batch = batch.append(doc)
    
    if len(batch) == batch_size:
        pending.append(asyncio.ensure_future(asyncio.ensure_future(_ainsert_single_batch(...))

    if len(pending) == max_concurrency:
        done, pending = await asyncio.wait(
            pending, return_when=asyncio.FIRST_COMPLETED
        )
await asyncio.wait(pending)
```